Src/Common/WebHeaderCollectionExtensions.cs

```diff
@@ -82,23 +121,84 @@ public static void SetNameValueHeaderValue(this NameValueCollection headers, str
                     if (keyValue.Length == 2 && keyValue[0].Trim() == keyName)
                     {
                         // Overwrite the existing thing
-                        headerValues[i] = string.Format(CultureInfo.InvariantCulture, "{0}={1}", keyName, value);
+                        headerValues[i] = FormatKeyValueHeader(keyName, value);
                         found = true;
                     }
                 }
 
-                headers[headerName] = string.Join(", ", headerValues);
+                headers[headerName] = string.Join(",", headerValues);
 
                 if (!found)
                 {
-                    headers[headerName] += string.Format(CultureInfo.InvariantCulture, ", {0}={1}", keyName, value);
+                    headers[headerName] += string.Format(CultureInfo.InvariantCulture, ",{0}={1}", keyName, value);
                 }
             }
             else
             {
                 // header with headerName does not exist - let's add one.
-                headers[headerName] = string.Format(CultureInfo.InvariantCulture, "{0}={1}", keyName, value);
+                headers[headerName] = FormatKeyValueHeader(keyName, value);
             }
         }
+
+        /// <summary>
+        /// For the given header collection, sets the header value based on the name value format.
+        /// </summary>
+        /// <param name="headers">Header collection.</param>
+        /// <param name="headerName">Name of the header that is to contain the name-value pair.</param>
+        /// <param name="keyValuePairs">List of KeyValuePairs to format into header.</param>
+        public static void SetHeaderFromNameValueCollection(this NameValueCollection headers, string headerName, IEnumerable<KeyValuePair<string, string>> keyValuePairs)
+        {
+            if (string.IsNullOrEmpty(headerName))
+            {
+                throw new ArgumentNullException(nameof(headerName));
+            }
+
+            var requiredHeader = headers[headerName];
+
+            // do not set header if it's present
+            if (string.IsNullOrEmpty(requiredHeader))
+            {
+                StringBuilder headerValue = new StringBuilder();
+                foreach (var pair in keyValuePairs)
+                {
+                    if (headerValue.Length > 0)
+                    {
+                        headerValue.Append(",");
+                    }
+
+                    headerValue.Append(FormatKeyValueHeader(pair.Key, pair.Value));
+                }
+
+                if (headerValue.Length > 0)
+                {
+                    headers[headerName] = headerValue.ToString();
+                }
+            }
+        }
+
+        private static bool TryParseKeyValueHeader(string pairString, out string key, out string value)
+        {
+            Debug.Assert(pairString != null, "pairString is null");
+            var keyValue = pairString.Split('=');
```

Comment:
```string.Split``` is expensive in memory allocations. For something as simple as this, you don't need it. The alternative doesn't look 'pretty', but I notice this is called in an inner loop, so it's probably worth optimizing. 
```C#
const char separator = '=';
int indexOfSeparator = pairString.IndexOf(separator, StringComparison.Ordinal);
// Must be a separator, and cannot be first or last
if (indexOfSeparator <= 0 || indexOfSeparator == pairString.Length -1) return false;
// Must be exactly one separator
if (pairString.IndexOf('=', indexOfSeparator + 1, StringComparison.Ordinal) >= 0) return false;
var key = pairString.Substring(0, indexOfSeparator).Trim();
var value = pairString.Substring(indexOfSeparator + 1).Trim();
if (key.Length > 0 && value.Length > 0)
{
    return true;
}

key = null;
value = null;
return false;